internal/exec/stages/files/files.go

```diff
@@ -79,57 +80,132 @@ func (s stage) Run(config types.Config) bool {
+type filesystemEntry interface {
+	create(l *log.Logger, u util.Util) error
+}
+
+type fileEntry types.File
+
+func (tmp fileEntry) create(l *log.Logger, u util.Util) error {
+	f := types.File(tmp)
+	file := util.RenderFile(l, f)
+	if file == nil {
+		return fmt.Errorf("failed to resolve file %q", f.Path)
+	}
+
+	if err := l.LogOp(
+		func() error { return u.WriteFile(file) },
+		"writing file %q", string(f.Path),
+	); err != nil {
+		return fmt.Errorf("failed to create file %q: %v", file.Path, err)
+	}
+
+	return nil
+}
+
+type dirEntry types.Directory
+
+func (tmp dirEntry) create(l *log.Logger, u util.Util) error {
+	d := types.Directory(tmp)
+	err := l.LogOp(func() error {
+		path := u.JoinPath(string(d.Path))
+		if err := util.MkdirForFile(path); err != nil {
+			return err
+		}
+
+		if err := os.Mkdir(path, os.FileMode(d.Mode)); err != nil {
+			return err
+		}
+
+		// Set the ownership and mode are as requested (since Mkdir doesn't set ownership and can be affect by the sticky bit)
+		if err := os.Chmod(path, os.FileMode(d.Mode)); err != nil {
+			return err
+		}
+
+		return os.Chown(path, d.User.Id, d.Group.Id)
+	}, "creating directory %q", string(d.Path))
+	if err != nil {
+		return fmt.Errorf("failed to create directory %q: %v", d.Path, err)
 	}
 
+	return nil
+}
+
+// ByDirectorySegments is used to sort directories so /foo get created before /foo/bar if they are both specified.
+type ByDirectorySegments []types.Directory
+
+func (lst ByDirectorySegments) Len() int { return len(lst) }
+
+func (lst ByDirectorySegments) Swap(i, j int) {
+	lst[i], lst[j] = lst[j], lst[i]
+}
+
+func (lst ByDirectorySegments) Less(i, j int) bool {
+	return lst[i].Depth() < lst[j].Depth()
+}
+
+// mapEntriesToFilesystems builds a map of filesystems to files. If multiple
+// definitions of the same filesystem are present, only the final definition is
+// used.
+func (s stage) mapEntriesToFilesystems(config types.Config) (map[types.Filesystem][]filesystemEntry, error) {
 	filesystems := map[string]types.Filesystem{}
 	for _, fs := range config.Storage.Filesystems {
 		filesystems[fs.Name] = fs
 	}
 
-	fileMap := map[types.Filesystem][]types.File{}
-	for fsn, f := range files {
-		if fs, ok := filesystems[fsn]; ok {
-			fileMap[fs] = append(fileMap[fs], f...)
+	entryMap := map[types.Filesystem][]filesystemEntry{}
+
+	// Sort directories to ensure /a gets created before /a/b.
+	sortedDirs := config.Storage.Directories
+	sort.Sort(ByDirectorySegments(sortedDirs))
+
+	// Add directories first to ensure they are created before files.
+	for _, d := range sortedDirs {
+		if fs, ok := filesystems[d.Filesystem]; ok {
+			entryMap[fs] = append(entryMap[fs], dirEntry(d))
 		} else {
-			s.Logger.Crit("the filesystem (%q), was not defined", fsn)
+			s.Logger.Crit("the filesystem (%q), was not defined", d.Filesystem)
 			return nil, ErrFilesystemUndefined
 		}
 	}
 
-	return fileMap, nil
+	for _, f := range config.Storage.Files {
```

Comment:
I don't think so without sacraficing type safety. Because you can't convert concrete types in go, even to/from an embedded struct, you can't write a generic "add `Node`-like object" function to add them to entryMap without it taking an interface{}. I figured the code duplication was worth the type safety. Additionally, you need the dual for loops anyway since you need to loop over config.Storage.{Files,Directories} and create either fileEntry or dirEntry depending on which one it is.